pkg/controller/clone-controller.go

```diff
@@ -0,0 +1,264 @@
+	
+)
+
+type CloneController struct {
+	clientset                kubernetes.Interface
+	pvcQueue, podQueue       workqueue.RateLimitingInterface
+	pvcInformer, podInformer cache.SharedIndexInformer
+	cloneImage               string
+	pullPolicy               string // Options: IfNotPresent, Always, or Never
+	verbose                  string // verbose levels: 1, 2, ...
+}
+
+func NewCloneController(client kubernetes.Interface, pvcInformer, podInformer cache.SharedIndexInformer, cloneImage string, pullPolicy string, verbose string) *CloneController {
+	c := &CloneController{
+		clientset:     client,
+		pvcQueue:      workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		podQueue:      workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		pvcInformer:   pvcInformer,
+		podInformer:   podInformer,
+		cloneImage:    cloneImage,
+		pullPolicy:    pullPolicy,
+		verbose:       verbose,
+	}
+
+	// Bind the pvc SharedIndexInformer to the pvc queue
+	c.pvcInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				c.pvcQueue.AddRateLimited(key)
+			}
+		},
+		// this is triggered by an update or it will also be
+		// be triggered periodically even if no changes were made.
+		UpdateFunc: func(old, new interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(new)
+			if err == nil {
+				c.pvcQueue.AddRateLimited(key)
+			}
+		},
+	})
+
+	// Bind the pod SharedIndexInformer to the pod queue
+	c.podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				c.podQueue.AddRateLimited(key)
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			
+			//TEMPORARY - for debug -  to be delete
+			//tmp, _ := json.Marshal(&newObj))
+		    //glog.V(Vdebug).Info(fmt.Println(string(tmp)))
+			///////////
+			
+			key, err := cache.MetaNamespaceKeyFunc(newObj)
+			if err == nil {
+				c.podQueue.AddRateLimited(key)
+			}
+		},
+	})
+
+	return c
+}
+
+func (c *CloneController) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer func() {
+		c.pvcQueue.ShutDown()
+		c.podQueue.ShutDown()
+	}()
+	glog.V(Vadmin).Infoln("Starting clone controller Run loop")
+	if threadiness < 1 {
+		return errors.Errorf("expected >0 threads, got %d", threadiness)
+	}
+
+	if !cache.WaitForCacheSync(stopCh, c.pvcInformer.HasSynced) {
+		return errors.New("Timeout waiting for pvc cache sync")
+	}
+	if !cache.WaitForCacheSync(stopCh, c.podInformer.HasSynced) {
+		return errors.New("Timeout waiting for pod cache sync")
+	}
+	glog.V(Vdebug).Infoln("CloneController cache has synced")
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runPVCWorkers, time.Second, stopCh)
+		go wait.Until(c.runPodWorkers, time.Second, stopCh)
+	}
+	<-stopCh
+	return nil
+}
+
+func (c *CloneController) runPodWorkers() {
```

Comment:
remove

Reply:
@davidvossel Is it reasonable to keep this pr mimicking the import controller and then change both in a new pr?